### PR TITLE
[REVIEW] Port nvstrings Miscellaneous functions to cuDF Python/Cython

### PR DESCRIPTION
**Pull request by galipremsagar**
This resolves: #3993 

Ported the following APIs:

~~- [ ] `nvstrings.nvstrings.compare` --> `Series.str.compare` (No built-in Pandas equivalent)~~ **Did not port this to python/cython as there is no port for this in strings libcudf**
- [x] `nvstrings.nvstrings.len` --> `Series.str.len` (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.len.html#pandas.Series.str.len)
- [x] `nvstrings.nvstrings.translate` --> `Series.str.translate` (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.translate.html#pandas.Series.str.translate)
- [x] `nvstrings.nvstrings.url_decode` --> `Series.str.url_decode` (No built-in Pandas equivalent)
- [x] `nvstrings.nvstrings.url_encode` --> `Series.str.url_encode` (No built-in Pandas equivalent)
- [x] `nvstrings.nvstrings.code_points` --> `Series.str.code_points` (No built-in Pandas equivalent)

- [x] Added function docs
- [x] Added unit-tests

### CODE REVIEW

File: python/cudf/cudf/_libxx/cpp/strings/attributes.pxd [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+from libcpp.memory cimport unique_ptr
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/cpp/strings/attributes.pxd]:
```suggestion

from libcpp.memory cimport unique_ptr
```

File: python/cudf/cudf/_libxx/cpp/strings/translate.pxd [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+from libcpp.memory cimport unique_ptr
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/cpp/strings/translate.pxd]:
```suggestion

from libcpp.memory cimport unique_ptr
```

File: python/cudf/cudf/_libxx/strings/translate.pyx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    mapping_table_dict = OrderedDict()
+
+    for key in mapping_table:
+        value = mapping_table[key]
+        if type(value) is str:
+            value = ord(value)
+        mapping_table_dict[(key, value)] = None
+
+    cdef vector[pair[char_utf8, char_utf8]] c_mapping_table
+    c_mapping_table = list(mapping_table_dict.keys())
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/strings/translate.pyx]: Cython handles translating a tuple into a pair without issue?

File: python/cudf/cudf/_libxx/strings/translate.pyx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    cdef unique_ptr[column] c_result
+    cdef column_view source_view = source_strings.view()
+
+    mapping_table_dict = OrderedDict()
+
+    for key in mapping_table:
+        value = mapping_table[key]
+        if type(value) is str:
+            value = ord(value)
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/strings/translate.pyx]:
```suggestion
        if type(value) is str:
            value = ord(value)
        if type(key) is str:
            key = ord(key)
```

File: python/cudf/cudf/_libxx/strings/translate.pyx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+def translate(Column source_strings,
+              object mapping_table):
+    """
+    Translates individual characters within each string
+    mapping present in the mapping_table.
+    """
+    cdef unique_ptr[column] c_result
+    cdef column_view source_view = source_strings.view()
+
+    mapping_table_dict = OrderedDict()
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/strings/translate.pyx]: Why use an `OrderedDict` here if you don't need to do lookups? Can just use a list. Or better yet, create the `c_mapping_table` vector before this loop, reserve the size needed, and `push_back` into the vector from this loop.

File: python/cudf/cudf/core/column/string.py [hunk 6]
```diff
@@ -1663,6 +1663,72 @@ def match(self, pat, case=True, flags=0, **kwargs):
...
+        This function uses the len() method to determine
+        the size of each sub-array of integers.
+
+        Returns
+        -------
+        Series or Index.
+        """
+        from cudf.core.series import Series
+
+        return Series(cpp_code_points(self._column),)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py]: Why can't we do `self._return_or_inplace` here? We shouldn't hardcode Series.

Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: `self._parent` and result from `code_points` are going to be of different lengths, thus conflicting with the length of index (`self._parent.index) we use in `_return_or_inplace`.

File: python/cudf/cudf/_libxx/strings/translate.pyx (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/_libxx/strings/translate.pyx]: Yep, it does.

File: python/cudf/cudf/core/column/string.py (hunk 6 above)
Comment (kkraus14) [reply python/cudf/cudf/core/column/string.py]: I see, so we should be able to do something like:
```
if isinstance(self._parent, Series):
    return Series(new_col, name=self._parent.name)
elif isinstance(self._parent, Index):
    return as_index(new_col, name=self._parent.index)
```

Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: Made changes

File: python/cudf/cudf/core/column/string.py [hunk 7]
```diff
@@ -1663,6 +1663,78 @@ def match(self, pat, case=True, flags=0, **kwargs):
...
+        """
+        from cudf.core.series import Series, Index
+
+        new_col = cpp_code_points(self._column)
+        if self._parent is None:
+            return new_col
+        elif isinstance(self._parent, Series):
+            return Series(new_col, name=self._parent.name)
+        elif isinstance(self._parent, Index):
+            return column.as_index(new_col, name=self._parent.index.name)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py]: `self._parent` is already an Index here
```suggestion
            return column.as_index(new_col, name=self._parent.name)
```

**Pull request merged**